[jax2tf] Added conversion for scatter*_p primitives.

bchetioui opened the pull request:
Limitations:

- the conversion works as well as the conversion of the underlying reduction functions (e.g. `lax.scatter_max` is not properly converted for the `int8` dtype, because `tf.math.maximum` is not defined for `int8` tensors);
- the conversion can not take advantage of the `unique_indices` parameter. This does not affect correctness, but may affect performance on certain platforms (as stated in the documentation of `lax.scatter`).

bchetioui committed:
[jax2tf] Added conversion for scatter*_p primitives.

gnecula commented on jax/experimental/jax2tf/tests/primitive_harness.py:
```diff
@@ -314,6 +314,46 @@ def parameterized(harness_group: Iterable[Harness],
+  # Directly from lax.scatter in tests/lax_test.py
+  Harness(
+    f"fun={f_lax.__name__}_shape={jtu.format_shape_dtype_string(shape, dtype)}_scatterindices={scatter_indices.tolist()}_updateshape={update_shape}_updatewindowdims={dimension_numbers.update_window_dims}_insertedwindowdims={dimension_numbers.inserted_window_dims}_scatterdimstooperanddims={dimension_numbers.scatter_dims_to_operand_dims}_indicesaresorted={indices_are_sorted}_uniqueindices={unique_i ...
+    (lambda f_lax: lambda *args: f_lax(*args[:-2], indices_are_sorted=args[-2], unique_indices=args[-1]))(f_lax),
```
I am wondering if this can be written more simply. What is the application to `f_lax` for? Can we just write `indices_are_sorted=indices_are_sorted` and not pass `StaticArg(indices_are_sorted)`?

SIben replied:
The application to `f_jax` is to avoid the problem described [here](https://docs.python.org/3.4/faq/programming.html#why-do-lambdas-defined-in-a-loop-with-different-values-all-return-the-same-result). Essentially, just writing `lambda *args: f_lax(*args[:-2], indices_are_sorted=args[-2], unique_indices=args[-1])` would result in `f_lax` always being `lax.scatter_add`.

Yes, we can set `indices_are_sorted=indices_are_sorted`, but then we also have to pass `indices_are_sorted` as a parameter to the outer lambda (and same for `unique_indices`, of course).

gnecula replied:
Do we even need a lambda? Why isn't `f_lax` enough?

SIben replied:
Because the functions are declared as `f(args, *, kwargs)`, the `kwargs` must be passed by name explicitly. I guess that another way to do it (that maybe is more elegant?) would be to remove the `StaticArg(indices_are_sorted)` and `StaticArg(unique_indices)` and replace this lambda by a partial application setting the `kwargs`?

SIben replied:
I addressed both your comments by replacing the call with a `partial` application in my latest commit. Let me know if that works :)

gnecula replied:
This is much nicer, thanks!

I am surprised that `dimension_numbers` does not need the same treatment.

SIben replied:
From the definition of `scatter_add`

```python
def scatter_add(operand: Array, scatter_indices: Array, updates: Array,
                dimension_numbers: ScatterDimensionNumbers, *,
                indices_are_sorted: bool = False,
                unique_indices: bool = False) -> Array:
```

you can see that `dimension_numbers` can be passed as a positional argument, so it does not pose the same issue.

gnecula commented on jax/experimental/jax2tf/jax2tf.py:
```diff
@@ -1120,31 +1120,41 @@ def _scatter_shape(operand, scatter_indices, updates, dimension_numbers):
-                    experimental_compile=True)(operand, scatter_indices, updates)
+    tf.function(update_computation).get_concrete_function(o_spec, o_spec))
+
+  out = tfxla.scatter(operand, scatter_indices, updates, xla_update_computation,
```
I would keep the `tf.function` wrapper for now, I suspect that this change will fail on GPU. I have PR #4030 to finish to clean this.

SIben replied:
Added back the wrapper in the latest commit.

bchetioui committed:
Put tf.function experimental compile wrapper back on scatter.

bchetioui committed:
Changed dirty lambda to partial application.

bchetioui committed:
Removed unique_indices=True test cases

bchetioui committed:
Remove non-deterministic test cases from the scatter harness.

This commit also documents the reasons for ignoring these test
cases and potential pitfalls, in case someone needs to perform
these tests at a later time.

bchetioui committed:
Added failure cases on TPU

bchetioui committed:
Removed double call to scatter.
